src/aggregator/integration/multi_server_forwarding_pipeline_test.go

```diff
@@ -339,6 +341,12 @@ func testMultiServerForwardingPipeline(t *testing.T, discardNaNAggregatedValues
 		time.Sleep(time.Second)
 	}
 
+	// Remove all the topic consumers before closing clients and servers. This allows to close the
+	// connections between servers while they still are running. Otherwise, during server shutdown,
+	// the yet-to-be-closed servers would repeatedly try to reconnect to recently closed ones, which
+	// results in longer shutdown times.
+	require.NoError(t, removeAllTopicConsumers(topicService, m3msgTopicName))
```

Comment:
The tests are way slower when using m3msg vs. TCP clients. The difference is especially large for the multi-server tests:
```
TCP:
PASS: TestMultiServerForwardingPipelineKeepNaNAggregatedValues (36.52s)
PASS: TestMultiServerForwardingPipelineDiscardNaNAggregatedValues (36.44s)

m3msg:
PASS: TestMultiServerForwardingPipelineKeepNaNAggregatedValues (240.36s)
PASS: TestMultiServerForwardingPipelineDiscardNaNAggregatedValues (246.05s)
```

Reply:
`240.36s` feels like too much. Do the logs help to understand which part of the test run is the biggest offender?

Reply:
Managed to cut some time by initializing single test client instead of one for each server (https://github.com/m3db/m3/pull/3681/commits/ca67b517735493ed423f59b71116aa90ecec7f59#diff-d3909b6c8294bf60de5bafd6ddd80c92ce0f534a7abd9b37e9438350ea34561eR203-R343). The tests now run faster, though it's still threefold slower than TCP ones.

```
TCP:
PASS: TestMultiServerForwardingPipelineKeepNaNAggregatedValues (36.42s)
PASS: TestMultiServerForwardingPipelineDiscardNaNAggregatedValues (36.43s)

m3msg:
PASS: TestMultiServerForwardingPipelineKeepNaNAggregatedValues (120.12s)
PASS: TestMultiServerForwardingPipelineDiscardNaNAggregatedValues (116.99s)
```

Reply:
Do you know what the minimum connection count is for m3msg for these clients used in the test? The default number of connections is 4 I think? We can probably tune this down for these tests since there is setup and teardown time used per connection which is considerable:
https://github.com/m3db/m3/blob/37c36bef78c552ae5b432aa35baa463752bbabbe/src/msg/producer/writer/options.go#L44-L52

If it's not this option we should also see what other things we may want to tweak.

Reply:
Unfortunately, reducing connection count didn't help, though I've found what was slowing things down.

Most of the slowdown was related to m3msg buffer's cleanup goroutine:
https://github.com/m3db/m3/blob/aa4c25a80a3427879896e4ec2ac9519b7c3bc200/src/msg/producer/buffer/buffer.go#L235-L236

Even though cleanup pass was quick, the retrier was sometimes taking more than a minute to return:
https://github.com/m3db/m3/blob/aa4c25a80a3427879896e4ec2ac9519b7c3bc200/src/msg/producer/buffer/buffer.go#L263-L273

After tuning retry options now the runtime is on par with TCP:
https://github.com/m3db/m3/pull/3681/commits/fbe7d3c2b61e77cc941080513b91f664911a1a3d

```
TCP:
PASS: TestMultiServerForwardingPipelineKeepNaNAggregatedValues (36.48s)
PASS: TestMultiServerForwardingPipelineDiscardNaNAggregatedValues (36.44s)

m3msg:
PASS: TestMultiServerForwardingPipelineKeepNaNAggregatedValues (38.02s)
PASS: TestMultiServerForwardingPipelineDiscardNaNAggregatedValues (39.14s)
```